File: firmware/FlockNoir/irsensor.cpp
#include "irsensor.h"
#include "pulse_detector.h"
#include <Preferences.h>
#include <esp_timer.h>
#include <esp32-hal-periman.h>

IrSensor irSensor;
static Preferences irPrefs;
void IrSensor::begin() {
  if (irPrefs.begin("flockir", true)) {
    _enabled = irPrefs.getBool("en", IR_DEFAULT_ENABLED != 0); irPrefs.end();
  }
  analogReadResolution(12);
  // The Arduino core attaches an ADC channel on its first read. Per-pin
  // attenuation only applies after that attachment has succeeded.
  analogRead(IR_SENSOR_PIN);
  if (perimanGetPinBusType(IR_SENSOR_PIN) != ESP32_BUS_TYPE_ADC_ONESHOT) {
    Serial.println("[IR] ADC initialization failed");
    return;
  }
  analogSetPinAttenuation(IR_SENSOR_PIN, ADC_11db);
  if (!_events) _events = xQueueCreate(8, sizeof(IrResult));
  if (!_events) Serial.println("[IR] event queue allocation failed");
  if (!_started) {
    _started = xTaskCreatePinnedToCore(taskThunk, "irsensor", 4096, this, 3, nullptr, 1) == pdPASS;
    if (!_started) Serial.println("[IR] task creation failed");
  }
}
bool IrSensor::enabled() const { return _enabled.load(); }
bool IrSensor::popEvent(IrResult &event) {
  while(_events && xQueueReceive(_events,&event,0)==pdTRUE)
    if(!suspended() && enabled() && event.generation==_generation.load())return true;
  return false;
}
void IrSensor::setSuspended(bool suspended) {
  if(_suspended.exchange(suspended)==suspended)return;
  ++_generation;
  portENTER_CRITICAL(&_mux);_res=IrResult();_res.present=_started;_rcount=_rhead=0;portEXIT_CRITICAL(&_mux);
}
void IrSensor::setEnabled(bool e) {
  ++_generation;
  _enabled.store(e);
  if (irPrefs.begin("flockir", false)) {
    if (!irPrefs.putBool("en", e)) Serial.println("[IR] settings write failed");
    irPrefs.end();
  }
}
IrResult IrSensor::result() {
  portENTER_CRITICAL(&_mux); IrResult r = _res; portEXIT_CRITICAL(&_mux);
  if (!enabled() || suspended()) r.detected = false;
  return r;
}
int IrSensor::snapshot(uint8_t *out, int maxN) {
  if (!out || maxN <= 0) return 0;
  uint16_t values[64];
  portENTER_CRITICAL(&_mux);
  int n = min(min(_rcount, maxN), 64);
  for (int i = 0; i < n; ++i) values[i] = _ring[(_rhead - n + i + IR_RING) % IR_RING];
  portEXIT_CRITICAL(&_mux);
  uint16_t low = 4095, high = 0;
  for (int i = 0; i < n; ++i) { low = min(low, values[i]); high = max(high, values[i]); }
  for (int i = 0; i < n; ++i) out[i] = (values[i] - low) * 100 / max(1, int(high - low));
  return n;
}
void IrSensor::taskThunk(void *arg) { static_cast<IrSensor *>(arg)->run(); }
void IrSensor::run() {
  PulseDetector pulse;
  pulse.config.minHz = IR_MIN_HZ; pulse.config.maxHz = IR_MAX_HZ;
  pulse.config.threshold = IR_THR_IDLE; pulse.config.release = IR_THR_LOCKED;
  pulse.config.required = IR_REQUIRED_INTERVALS;
  pulse.config.staleUs = IR_ACTIVE_WINDOW_MS * 1000;
  pulse.config.minDuty = IR_DUTY_MIN; pulse.config.maxDuty = IR_DUTY_MAX;
  pulse.config.minWidthMs = IR_PULSE_MIN_MS; pulse.config.maxWidthMs = IR_PULSE_MAX_MS;
  pulse.config.maxGapUs = IR_MAX_SAMPLE_GAP_US;
  TickType_t last = xTaskGetTickCount();
  const TickType_t ticks = max(TickType_t(1), pdMS_TO_TICKS(1000 / IR_SAMPLE_HZ));
  uint32_t publish = 0, samples = 0, rateAt = millis();
  float rate = 0;
  int decim = 0;
  bool wasEnabled = false;uint32_t generation=_generation.load();
  uint32_t lastEvent = 0, droppedEvents = 0;
  bool previousMatch = false;
  for (;;) {
    if(suspended()) {
      pulse.reset();wasEnabled=false;previousMatch=false;rate=0;samples=0;rateAt=millis();
      vTaskDelay(pdMS_TO_TICKS(20));last=xTaskGetTickCount();continue;
    }
    uint32_t currentGeneration=_generation.load();
    if(generation!=currentGeneration){pulse.reset();generation=currentGeneration;previousMatch=false;}
    bool en = enabled();
    uint16_t raw = analogRead(IR_SENSOR_PIN);
    uint32_t now = millis();
    if (en != wasEnabled) { pulse.reset(); wasEnabled = en; previousMatch = false; }
    if (en) pulse.feed(raw, uint32_t(esp_timer_get_time()));
    ++samples;
    if (now - rateAt >= 1000) { rate = samples * 1000.0f / (now - rateAt); samples = 0; rateAt = now; }
    if (++decim >= max(1, IR_SAMPLE_HZ / 200)) {
      decim = 0;
      portENTER_CRITICAL(&_mux);
      _ring[_rhead] = raw; _rhead = (_rhead + 1) % IR_RING;
      if (_rcount < IR_RING) ++_rcount;
      portEXIT_CRITICAL(&_mux);
    }
    if (now - publish >= 100) {
      publish = now;
      IrResult r;
      r.detected = en && pulse.matched;
      r.freqHz = pulse.frequency; r.dutyCycle = pulse.duty;
      r.amp = uint16_t(pulse.amplitude); r.baseline = uint16_t(pulse.baseline);
      r.validCount = pulse.valid; r.present = _started;
      r.clipped = pulse.clipped; r.pulseMs = pulse.widthMs;
      r.sampleHz = rate; r.gaps = pulse.gaps; r.noise = pulse.noise; r.raw = raw;
      r.timestampMs = now;r.generation=generation;
      if (r.detected && (!previousMatch || now-lastEvent >= ALERT_HOLDOFF_MS)) {
        if (!_events || xQueueSend(_events,&r,0)!=pdTRUE) ++droppedEvents;
        lastEvent = now;
      }
      previousMatch = r.detected;
      r.droppedEvents = droppedEvents;
      portENTER_CRITICAL(&_mux); if(!suspended() && generation==_generation.load())_res = r; portEXIT_CRITICAL(&_mux);
    }
    vTaskDelayUntil(&last, ticks);
    if (xTaskGetTickCount() - last > ticks) last = xTaskGetTickCount();
  }
}

File: pi/flocknoir/irsensor.py
"""OPT101 -> MCP3008 sampler using shared validation and latched events.

Linux scheduling is not hard real time. Rate/gap diagnostics expose timing
failures. ADC activity cannot prove an optical sensor is connected.
"""
from collections import deque
import queue
import threading
import time

import config as C
from . import settings
from .native import Pulse


class IrSensor:
    def __init__(self, start=True):
        self._lock = threading.Lock()
        self._stop = threading.Event()
        self.enabled = bool(settings.load_section("irsensor", {}).get("enabled", C.IR_ENABLED_DEFAULT))
        self.events = queue.Queue(maxsize=32)
        self._ring = deque(maxlen=256)
        self.res = dict(detected=False, freqHz=0., dutyCycle=0., amp=0, baseline=0,
                        validCount=0, present=False, clipped=False, pulseMs=0., sampleHz=0.,
                        gaps=0, droppedEvents=0, noise=0., raw=0, timestamp=0.)
        self.error = ""
        if start:
            threading.Thread(target=self._run, daemon=True, name="opt101").start()

    def set_enabled(self, enabled):
        settings.save_section("irsensor", {"enabled": bool(enabled)})
        with self._lock:
            self.enabled = bool(enabled)
            if not self.enabled:
                self.res["detected"] = False

    def result(self):
        with self._lock:
            result = dict(self.res)
            if not self.enabled or time.monotonic() - result["timestamp"] > 0.3:
                result["detected"] = False
            return result

    def pop_event(self, timeout=0.2):
        try:
            return self.events.get(timeout=timeout)
        except queue.Empty:
            return None

    def snapshot(self, max_n=64):
        with self._lock:
            values = list(self._ring)[-max_n:]
        if not values:
            return []
        low, high = min(values), max(values)
        return [int((v-low)*100/max(1, high-low)) for v in values]

    def stop(self):
        self._stop.set()

    def _run(self):
        while not self._stop.is_set():
            adc = pulse = None
            try:
                from gpiozero import MCP3008
                pulse = Pulse(
                    [C.IR_MIN_HZ, C.IR_MAX_HZ, C.IR_DUTY_MIN, C.IR_DUTY_MAX,
                     C.IR_PULSE_MIN_MS, C.IR_PULSE_MAX_MS, C.IR_THR_IDLE, C.IR_THR_LOCKED, 6],
                    [C.IR_REQUIRED_INTERVALS, C.IR_ACTIVE_WINDOW_MS*1000, C.IR_MAX_SAMPLE_GAP_US])
                adc = MCP3008(channel=C.IR_ADC_CHANNEL, port=C.IR_SPI_BUS, device=C.IR_SPI_DEVICE)
                self.error = ""
                self._sample(adc, pulse)
            except Exception as exc:
                self.error = str(exc)
                with self._lock:
                    self.res.update(present=False, detected=False)
                self._stop.wait(5)
            finally:
                if adc is not None:
                    adc.close()
                if pulse is not None:
                    pulse.close()

    def _sample(self, adc, pulse):
        interval = 1 / C.IR_SAMPLE_HZ
        next_sample = rate_at = time.monotonic()
        publish = last_event = 0.
        rate = 0.
        samples = decimation = dropped = 0
        was_enabled = previous_match = False
        while not self._stop.is_set():
            raw = round(adc.value * 4095)
            now = time.monotonic()
            enabled = self.enabled
            if enabled != was_enabled:
                pulse.reset()
                was_enabled, previous_match = enabled, False
            if enabled:
                pulse.feed(raw, int(now * 1_000_000))
            samples += 1
            if now-rate_at >= 1:
                rate = samples / (now-rate_at)
                samples, rate_at = 0, now
            decimation += 1
            if decimation >= max(1, C.IR_SAMPLE_HZ // 200):
                decimation = 0
                with self._lock:
                    self._ring.append(raw)
            if now-publish >= 0.1:
                publish = now
                result = pulse.result()
                result.update(detected=enabled and result["detected"], present=True,
                              sampleHz=rate, raw=raw, timestamp=now)
                if result["detected"] and (not previous_match or (now-last_event)*1000 >= C.ALERT_HOLDOFF_MS):
                    try:
                        self.events.put_nowait(dict(result))
                    except queue.Full:
                        dropped += 1
                    last_event = now
                previous_match = result["detected"]
                result["droppedEvents"] = dropped
                with self._lock:
                    self.res = result
            next_sample += interval
            wait = next_sample-time.monotonic()
            if wait > 0:
                self._stop.wait(wait)
            else:
                next_sample = time.monotonic()
